### DISCUSSION TITLE
Can I apply a prefix to all keys loaded from a configuration provider?

### DISCUSSION THREAD

**Discussion by wazzamatazz** [dotnet/runtime#60458] [Q&A]
Say I have an options class like this:

```csharp
public class MyOptions {
  public string Item1 { get; set; }
  public string Item2 { get; set; }
  public IDictionary<string, Device> Devices { get; set; }
}

public class Device {
  public string DisplayName { get; set; }
  public string ConnectionString { get; set; }
  public IEnumerable<string> SubscriptionItems { get; set; }
}
```

In my startup code, I might bind these options as follows:

```csharp
services.Configure<MyOptions>(Configuration.GetSection("MyApp:MyOptions"));
```

If I wanted to define individual items in the `MyOptions.Devices` collection in separate files, I could add additional JSON providers to my `IConfigurationBuilder` and then define additional configuration files e.g.:

```json
// devices.sensor-01.json
{
  "MyApp": {
    "MyOptions": {
      "Devices": {
        "sensor-01": {
          "DisplayName": "Test sensor 1",
          "ConnectionString": "SOME_CONNECTION",
          "SubscriptionItems": [
            "Temperature-01",
            "Temperature-02"
          ]
        }
      }
    }
  }
}
```

```json
// devices.sensor-02.json
{
  "MyApp": {
    "MyOptions": {
      "Devices": {
        "sensor-02": {
          "DisplayName": "Test sensor 2",
          "ConnectionString": "SOME_CONNECTION",
          "SubscriptionItems": [
            "Temperature-98",
            "Temperature-99"
          ]
        }
      }
    }
  }
}
```

This works well, but I'd like to know if there is a built-in way to automatically add a prefix to all keys loaded from a given provider? In my example above, I'd like to be able to prefix all keys from JSON files matching `devices.*.json` with `MyApp:MyOptions:Devices:`, so that the JSON file itself could be simplified to something like this:

```json
// devices.sensor-01.json (simplified)
{
  "sensor-01": {
    "DisplayName": "Test sensor 1",
    "ConnectionString": "SOME_CONNECTION",
    "SubscriptionItems": [
      "Temperature-01",
      "Temperature-02"
    ]
  }
}
```

Is this something that is possible out of the box, or would I have to write a custom JSON configuration provider to accomplish this? Looking at the source for the [JsonConfigurationFileParser](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Json/src/JsonConfigurationFileParser.cs), this could be accomplished by injecting an initial context into the parser containing the prefix, but since the class itself is `internal` and the `EnterContext` method is `private`, there's no obvious way to accomplish it without copying and modifying this code, unless it is possible to tell the configuration builder itself to apply a prefix to keys from certain sources?

**Comment by FiniteReality** [comment]
I've worked around this before using the IEnumerable overload in the configuration provider; you build one configuration root with the values you want, and then enumerate its key/value pairs, adding a prefix to the key. Something like this:
```cs
var rootOne = new ConfigurationBuilder()
    .AddYourSource()
    .Build();

realRoot.AddInMemoryCollection(AddPrefix(rootOne.AsEnumerable(), "prefix"));

static IEnumerable<KeyValuePair<string, string>> AddPrefix(IEnumerable<KeyValuePair<string, string>> values, string prefix)
{
    foreach (var pair in values)
        yield return new(ConfigurationPath.Combine(prefix, pair.Key), pair.Value);
}
```

Unfortunately, this does not allow for things like reload tokens. I was looking into something like this at one point, but I couldn't figure out a good way to implement that without being confusing.

**Reply by wazzamatazz** [reply]
Thanks for the suggestion. Being able to reload on change would probably be preferable in the use cases I'm thinking about, unfortunately!
